tokio/src/util/wake_list.rs

```diff
@@ -38,13 +40,14 @@ impl WakeList {
 
     pub(crate) fn wake_all(&mut self) {
         assert!(self.curr <= NUM_WAKERS);
-        while self.curr > 0 {
-            self.curr -= 1;
-            // SAFETY: The first `curr` elements of `WakeList` are initialized, so by decrementing
-            // `curr`, we can take ownership of the last item.
-            let waker = unsafe { ptr::read(self.inner[self.curr].as_mut_ptr()) };
+        for i in 0..self.curr {
+            // SAFETY: The first `curr` elements are initialized
+            let waker = unsafe {
+                ptr::replace(self.inner.as_mut_ptr().add(i).cast::<Waker>(), noop_waker())
+            };
             waker.wake();
         }
+        self.curr = 0;
```

Comment:
We don't need to introduce a whole noop waker just to make this panic safe. We can use a helper struct with a drop implementation. For example:

1. Make a struct that keeps a start and end index of the remaining wakers.
2. When calling `wake_all`, immediately set `self.curr` to zero, and make the mentioned struct with the full range.
3. As long as the start and end indexes are different, you loop: Increment the start index by one, afterwards do the `ptr::read` + `waker.wake()` dance on the waker you just removed from the range.
4. On drop of the helper struct, call `drop_in_place` on the remaining slice.